test/e2e/simple_test.go

```diff
@@ -38,45 +37,30 @@ import (
 // the test namespace.
 func TestMain(m *testing.M) {
 	flag.Parse()
-	clients, err := newClients(test.Flags.Kubeconfig, test.Flags.Cluster, buildTestNamespace)
-	if err != nil {
-		log.Fatalf("newClients: %v", err)
-	}
+	logging.InitializeLogger(test.Flags.LogVerbose)
+	logger := logging.GetContextLogger("TestSetup")
 
-	// Ensure the test namespace exists, by trying to create it and ignoring
-	// already-exists errors.
-	if _, err := clients.kubeClient.Kube.CoreV1().Namespaces().Create(&corev1.Namespace{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: buildTestNamespace,
-		},
-	}); err == nil {
-		log.Printf("Created namespace %q", buildTestNamespace)
-	} else if kuberrors.IsAlreadyExists(err) {
-		log.Printf("Namespace %q already exists", buildTestNamespace)
-	} else {
-		log.Fatalf("Error creating namespace %q: %v", buildTestNamespace, err)
-	}
+	clients := setup(logger)
+	test.CleanupOnInterrupt(func() { teardownNamespace(clients, logger) }, logger)
 
+	code := m.Run()
 	defer func() {
+		// Cleanup namespace
```

Comment:
It doesn't add much to call `defer` at the end of a function, since there's nothing to defer from. If `Run` panics, we won't yet know to defer teardown. I think this should maybe be:

```
...
defer teardownNamespace(clients, logger)
os.Exit(m.Run())
```


```

Reply:
If `os.Exit` is invoked then `defer` is not respected.
Ref: https://gobyexample.com/exit

Reply:
In 2nd option I need to declare `code` var before defer so I will try 1st option without `defer`.

Reply:
Actually thinking about it more, the first option will not `teardownNamespaces` if `m.Run` errors out, but if you have the defer before, it will be cleaned up

Reply:
`m.Run` should return non zero exit code if it errors but if `m.Run` has any non recoverable errors then defer might good way to catch it all.